core/src/main/java/org/elasticsearch/discovery/zen/PublishClusterStateAction.java

```diff
@@ -90,6 +98,10 @@
     private final IncomingClusterStateListener incomingClusterStateListener;
     private final DiscoverySettings discoverySettings;
 
+    private AtomicLong fullClusterStateReceivedCount = new AtomicLong();
+    private AtomicLong clusterStateDiffReceivedCount = new AtomicLong();
+    private AtomicLong compatibleClusterStateDiffReceivedCount = new AtomicLong();
```

Comment:
I personally find having a incompatibleDiff count more handy as it's what we'll be looking for. Maybe just have compatibleDiffCount and incompatibleDiffCount (instead of total)?

Reply:
Ok. How picky do we want to be about concurrency here? I can do this by counting the cluster state diff as incompatible until successfully applied (i.e. no exceptions thrown) and then to decrement the `incompatible` count and increment the `compatible` one, but that does mean you could see incorrect values if you get the stats at the wrong moment.

On second thought the `stats()` method is already a bit racy.

I have various other, more complicated, ideas:
- counting more things, to distinguish in-progress applications from rejected ones.
- catching an `IncompatibleClusterStateVersionException` on the way past and updating the stats in the handler.
- keeping the accumulators as-is and calculating the difference in the `stats()` method (potentially also racy, not sure yet).
- introducing locks around stats reads and updates.

Any particular preferences?